### ISSUE TITLE
Add mode='valid' to scipy.ndimage.convolve and scipy.ndimage.convolve1d

### ISSUE THREAD

**Issue by carlosgmartin** [scipy/scipy#12997]
Add `mode='valid'` to [`scipy.ndimage.convolve`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.convolve.html) and [`scipy.ndimage.convolve1d`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.convolve1d.html), as exists for [`numpy.convolve`](https://numpy.org/doc/stable/reference/generated/numpy.convolve.html). This allows users to keep only the valid parts of the convolution (that do not interact with boundaries).

**Comment by huuanhhuynguyen** [comment]
Hi, I could implement it. Do we need this change?

**Comment by grlee77** [comment]
> Hi, I could implement it. Do we need this change?

`scipy.signal.convolve` and `scipy.signal.correlate` are n-dimensional and already support `'valid'`, `'same'` and `'full'` modes, so I don't know that it adds much value to also have this capability in `ndimage`. Those signal functions do not have a 1d equivalent, but you should be able to do 1d filtering with them by just making the filter size 1 on the dimensions that do not need to be filtered.

Would those existing functions meet your needs? Maybe we can add a note and see also to these in the docstrings of the ndimage functions to make them more discoverable. Being located in the `signal` module, I suspect many users may assume these would be 1d only.

**Comment by carlosgmartin** [comment]
@grlee77 I think it'd be more user-friendly to simply add this mode to `convolve` and `convolve1d`. This saves users the trouble of having to reimplement it every time in terms of other functions, which might also be error-prone.

**Comment by carlosgmartin** [comment]
@grlee77 The issue with [`scipy.signal.convolve`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.convolve.html) is that it doesn't let you convolve along a single axis, e.g. when you want to compute several [moving averages](https://stackoverflow.com/a/43200476).

**Comment by carlosgmartin** [comment]
@huuanhhuynguyen Here's an example for [`uniform_filter1d`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.uniform_filter1d.html):

```python3
from scipy.ndimage import uniform_filter1d

def filter(x, n, axis):
    x = np.asarray(x, dtype=float)
    slices = [None] * x.ndim
    slices[axis] = slice(n//2, 1 - n//2)
    return uniform_filter1d(x, n, axis, mode='constant')[tuple(slices)]
```

**Comment by grlee77** [comment]
> @grlee77 The issue with scipy.signal.convolve is that it doesn't let you convolve along a single axis, e.g. when you want to compute several moving averages.

Consider a batch of signals of where `signals.shape` is `(n, b)` where `n` is the signal length and `b` is the batch size. If you want to perform a moving average for a window of size 5 on all signals, my suggestion above was to use a kernel that is size 1 along the dimensions you don't want to average along. e.g.
```
scipy.signal.convolve(signals, np.ones((5, 1)), mode='valid')
```

I am not against the suggestion of having these modes supported in `scipy.ndimage` out of convenience, though. Just implementing this on the python side by truncating the output to the valid region should be fairly easy. Similarly, `'full'` mode can be implemented by prepadding the input with zeros.

Slicing the output involves performing computations in edge regions that end up getting discarded, but in cases with images that are large relative to the convolution kernels, the overhead may be minimal.  Avoiding computing the edge regions in the first place would require changes to the underlying C code, but that could always be done in a follow up PR if anyone is motivated. 

From an API standpoint, we cannot just use `mode='valid'` since the `mode` kwarg already exists in `scipy.ndimage` to represent the boundary extension mode. With a new keyword-only argument with a name like `extent`, `convolution_extent` or something along those lines, it could work, though. Any good name ideas?

**Comment by carlosgmartin** [comment]
@grlee77 Apologies if I misunderstood, but isn't `mode='valid'` already used by e.g. [`scipy.signal.convolve`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.convolve.html) for the same purpose?

If a different name is needed, maybe `'inner'`?

**Comment by carlosgmartin** [comment]
@grlee77 Ah. I see that `mode` has different meanings for [`scipy.signal`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.convolve.html) and [`scipy.ndimage`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.convolve.html). (The output shape doesn't change for the latter.) Is that what you mean?

Still, perhaps this could be considered a pseudo-extension mode? If not, maybe an `extent` keyword would be fine.

**Comment by grlee77** [comment]
> @grlee77 Ah. I see that mode has different meanings for scipy.signal and scipy.ndimage. (The output shape doesn't change across modes for the latter.) Is that what you mean?

Yes, in ndimage the function operates as for `mode='same'` from scipy.signal, but with several different options of how to treat the values outside the original array boundaries (e.g. periodic, symmetric, etc.). I suppose we could just add `mode='valid'` and mode='full' to the list and just set the boundary mode to `constant` with `cval=0` internally in those two cases.

Specifically the docstring for mode would look something like the following (would this seem reasonably clear to you as a user?)

```
"""mode : {'reflect', 'constant', 'nearest', 'mirror', 'wrap', 'valid', 'full'}, optional
    The `mode` parameter determines how the input array is extended
    beyond its boundaries. Default is 'reflect'. Behavior for each valid
    value is as follows:

    'reflect' (`d c b a | a b c d | d c b a`)
        The input is extended by reflecting about the edge of the last
        pixel. This mode is also sometimes referred to as half-sample
        symmetric.

    'constant' (`k k k k | a b c d | k k k k`)
        The input is extended by filling all values beyond the edge with
        the same constant value, defined by the `cval` parameter.

    'nearest' (`a a a a | a b c d | d d d d`)
        The input is extended by replicating the last pixel.

    'mirror' (`d c b | a b c d | c b a`)
        The input is extended by reflecting about the center of the last
        pixel. This mode is also sometimes referred to as whole-sample
        symmetric.

    'wrap' (`a b c d | a b c d | a b c d`)
        The input is extended by wrapping around to the opposite edge.

    'valid'
        Any pixels that would involve overlap between the `weights` array and
        values outside the original image extent are discarded. The output will
        be smaller than the input.

    'full'
        This mode returns an array corresponding to the full discrete linear
        convolution of the inputs. Values outside of the boundaries of `input`
        and `weights` are assumed to be 0. The output will have a size larger
        than the input (specifically ``output.shape[i]`` will be
        ``input.shape[i] + weights.shape[i] - 1``).

    For consistency with the interpolation functions, the following mode
    names can also be used:

    'grid-mirror'
        This is a synonym for 'reflect'.

    'grid-constant'
        This is a synonym for 'constant'.

    'grid-wrap'
        This is a synonym for 'wrap'."""
```

**Comment by carlosgmartin** [comment]
@grlee77 That seems reasonably clear to me.

**Comment by CamiloMartinezM** [comment]
I suppose this wasn't implemented in the end? I just stumbled upon the need to use `convolve1d` with `axis` but having some trouble specifying a `mode`.

**Comment by alec-flowers** [comment]
Hi! I have another use case where this functionality would be appreciated.

**Comment by yuanx749** [comment]
It would be nice to have `valid` mode in `ndimage.convolve1d` and `ndimage.correlate1d`. In some cases they are faster than their counterparts in `signal` module.

**Comment by LucasRr** [comment]
Most filtering functions (`convolve()`, `correlate()`, `gaussian_filter()`, `generic_filter()`, `laplace()`, ...) and morphology functions (`grey_dilation()`, `grey_erosion()`, `grey_closing()`, ...) in `ndimage` would benefit from a `'valid'` mode. Most of them do not have an equivalent in `scipy.signal`, so valid filtering is not possible. When the kernel size is relatively large, doing a `'full'` filtering then cropping is highly inefficient. Allowing `'valid'` filtering for all `ndimage` functions would benefit a lot of people I'm sure.
